### PR TITLE
Implement reporting of errors to GitOpsDeployment CR status conditions

### PR DESCRIPTION
**Pull request by drpaneas**
Fixes https://issues.redhat.com/browse/GITOPSRVCE-45

Signed-off-by: Panagiotis Georgiadis <pgeorgia@redhat.com>

I would recommend you to enable the beta Tree feature of github so you can see the overview of the file structure while reviewing this PR, as it will give you a much clearer 1000 ft view of the directory structure and its changes.

1. At [backend/apis/managed-gitops/v1alpha1/gitopsdeployment_types.go](https://github.com/redhat-appstudio/managed-gitops/pull/76/files#diff-c1cfffcad58ef3ab259f139d18a20a388a36b3bc137d80d3b4941b9fe2c376dc) you will see changes to the API, mostly comments and renames. The important thing here is the addition of `LastProbeTime` which was missing.
2. The `backend/apis/managed-gitops/v1alpha1/mocks` folder is new. It has mocks for the k8s client, generated via mockgen from the upstream and also it has a sub-dir, called structs. Inside of the structs you will find helper mocks for testing the gitopsdeployment API.
3. A new package [backend/condition/conditions.go](https://github.com/redhat-appstudio/managed-gitops/pull/76/files#diff-9cf59d7c68f4219feab8a9e7c444b8db1f34bb8f1dca442d16ca46fd611dc837) that introduces the Conditions interface with its methods (SetCondition, FindCondition and HasCondition). In the same pkg you will find mocks (inside the mocks sub-dir) (again generated with mockgen) and unit-tests using Ginkgo.
4. The [backend/eventloop/application_event_runner.go](https://github.com/redhat-appstudio/managed-gitops/pull/76/files#diff-de5c9757e496bcef8038e5e6357e82e43f317a0c90ab6bee5763d64882792e91) takes action and calls the Condition methods. The actual logic of what is happening when an error is not `nil` can be found at `SetGitOpsDeploymentCondition()` function. I also added unit-tests in Ginkgo using the mocks.

### CODE REVIEW

**Review by jgwest — CHANGES_REQUESTED** [review]
Looks great! A couple questions, a minor tweak, then good to go.

File: backend/condition/conditions_test.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
+package condition
+
+import (
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	gitopsv1alpha1 "github.com/redhat-appstudio/managed-gitops/backend/apis/managed-gitops/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var _ = Describe("ConditionManager", func() {
```
[the remark this answers is not in the document]
Comment (drpaneas) [reply backend/condition/conditions_test.go:10]: I think I've brought this topic in the past, but I don't remember if there was any clear conclusion if we should use or not Ginkgo. Personally, I prefer Ginkgo mainly for two reasons:

1. It's widely used in Kubernetes upstream
2. It makes reading testing code clearer and simpler (for me)

File: backend/eventloop/application_event_runner_test.go [hunk 2]
```diff
@@ -289,3 +296,72 @@ func TestApplicationEventLoopRunner_handleSyncRunModified(t *testing.T) {
...
+		mockStatusWriter = mocks.NewMockStatusWriter(mockCtrl)
+	})
+	JustBeforeEach(func() {
+		adapter = NewGitOpsDeploymentAdapter(gitopsDeployment, log.Log.WithName("Test Logger"), mockClient, mockConditions, ctx)
+	})
+	AfterEach(func() {
+		mockCtrl.Finish()
+	})
+
+	Context("SetProjectClaimCondition()", func() {
```
[the remark this answers is not in the document]
Comment (drpaneas) [reply backend/eventloop/application_event_runner_test.go:325]: Ah yes, part of the code was copied from OpenShift's GCP Project Operator which I wrote a couple of years a back. Let me fix that asap

**Pull request merged**